shadow-dom/focus/focus-method-delegatesFocus.html

```diff
@@ -0,0 +1,289 @@
+shadowRoot.appendChild(firstSlot);
+shadowRoot.appendChild(secondSlot);
+shadowRoot.appendChild(belowSlots);
+
+const elementsInFlatTreeOrder = [host, aboveSlots, firstSlot,
+  slottedToFirstSlot, secondSlot, slottedToSecondSlot, belowSlots, outside];
+
+// Final structure:
+// <div #host> (delegatesFocus=true)
+//    #shadowRoot
+//      <div #aboveSlots>
+//      <slot #firstSlot>
+//        (slotted) <div #slottedToFirstSlot>
+//      <slot #secondSlot>
+//        (slotted) <div #slottedToSecondSlot>
+//      <div #belowSlots>
+// <div #outside>
+
+function setTabIndex(elements, value) {
+  for (const el of elements) {
+    el.tabIndex = value;
+  }
+}
+
+function setAllTabIndex(value) {
+  setTabIndex(elementsInFlatTreeOrder, value);
+}
+
+function removeAllTabIndex() {
+  for (const el of elementsInFlatTreeOrder) {
+    el.removeAttribute("tabindex");
+  }
+}
+
+function resetFocus() {
+  removeAllTabIndex();
+  if (shadowRoot.activeElement)
+    shadowRoot.activeElement.blur();
+  if (document.activeElement)
+    document.activeElement.blur();
+}
+
+test(() => {
+  resetFocus();
+  setAllTabIndex(0);
+  // Structure:
+  // <div #host> (delegatesFocus=true) tabindex=0
+  //    #shadowRoot
+  //      <div #aboveSlots> tabindex=0
+  //      <slot #firstSlot> tabindex=0
+  //        (slotted) <div #slottedToFirstSlot> tabindex=0
+  //      <slot #secondSlot> tabindex=0
+  //        (slotted) <div #slottedToSecondSlot> tabindex=0
+  //      <div #belowSlots> tabindex=0
+  // <div #outside> tabindex=0
+  // flattened tabindex-ordered focus navigation scope of #host:
+  // [aboveSlots, slottedToFirstSlot, slottedToSecondSlot, belowSlots]
+  host.focus();
+  assert_equals(shadowRoot.activeElement, aboveSlots);
+  assert_equals(document.activeElement, host);
+}, "focus() on host with delegatesFocus, all tabindex=0");
+
+test(() => {
+  resetFocus();
+  setAllTabIndex(0);
+  setTabIndex([host], -1);
+  // flattened tabindex-ordered focus navigation scope of #host:
+  // [aboveSlots, slottedToFirstSlot, slottedToSecondSlot, belowSlots]
+  host.focus();
+  assert_equals(shadowRoot.activeElement, aboveSlots);
+  assert_equals(document.activeElement, host);
+}, "focus() on host with delegatesFocus & tabindex =-1, all other tabindex=0");
+
+test(() => {
+  resetFocus();
+  setTabIndex([aboveSlots, slottedToFirstSlot, slottedToSecondSlot, belowSlots], 0);
+  // flattened tabindex-ordered focus navigation scope of #host:
+  // [aboveSlots, slottedToFirstSlot, slottedToSecondSlot, belowSlots]
+  host.focus();
+  assert_equals(shadowRoot.activeElement, aboveSlots);
+  assert_equals(document.activeElement, host);
+}, "focus() on host with delegatesFocus & no tabindex, all other tabindex=0");
+
+test(() => {
+  resetFocus();
+  setAllTabIndex(-1);
+  setTabIndex([host], 0);
+  // flattened tabindex-ordered focus navigation scope of #host: []
+  host.focus();
+  assert_equals(shadowRoot.activeElement, null);
+  assert_equals(document.activeElement, null);
```

Comment:
How does this expectation make sense? `resetFocus()` call above just moved the focus to `document.body`. There is no focusable area within this shadow tree so we should be bailing out before moving any focus so `host.focus()` should really be a no-op.

Also, the current definition makes it so that `document.activeElement` almost never returns `null` so long as the document has body or a document element:
https://html.spec.whatwg.org/multipage/interaction.html#dom-document-activeelement

Reply:
So I think you'd have to replace everywhere you assert `document.activeElement` to be `null` to assert it's `document.body`. Then all test cases would pass in WebKit once I've implemented all the relevant bits and pieces.

Reply:
`resetFocus()` calls `document.activeElement.blur()` which I assume would remove any existing focus in the document? At least in Blink that means `document.activeElement` would be null after. Is it not supposed to be like that?

Reply:
That's odd. The version of `shadow-utils.js` I have is as follows:
```js
function resetFocus() {
  document.body.focus();
}
```

Reply:
Aah I see, yeah this test file doesn't use `shadow-utils.js` at all. It uses the resetFocus() defined in line 67. Since the PR for the sequential focus WPT is merged already https://github.com/web-platform-tests/wpt/pull/17523, should I make these tests use shadow-utils as well? (or maybe I should just rename shadow-utils to sequential-focus-utils..)

Reply:
Updated the tests, PTAL - also added sequential focus navigation tests.